File: Dice/RxWp7Dice/Gestures/ShakeObserver.vb
Imports Microsoft.Phone.Reactive
Imports Microsoft.Devices.Sensors

Public Module ShakeObserver
    Const MinimumOffset = 1.44
    Const TimeThreshold = 200

    Public Function GetShakeObserver(ByVal accel As Accelerometer) As IObservable(Of Unit)
        Dim query = From start In Observable.FromEvent(Of AccelerometerReadingEventArgs)(accel, "ReadingChanged")
                    Where (start.EventArgs.X ^ 2 + start.EventArgs.Y ^ 2) > MinimumOffset
                    Select start

        Dim shake = From interval In query.TimeInterval
                    Where interval.Interval.TotalMilliseconds < MinimumOffset
                    Select New Unit

        Return shake
    End Function

    'Public Function GetShakeObserver(ByVal accel As Accelerometer) As IObservable(Of Unit)
    '    Dim query =
    '        From interval In
    '        (From startEvent In Observable.FromEvent(Of AccelerometerReadingEventArgs)(accel, "ReadingChanged")
    '          Where (startEvent.EventArgs.X * startEvent.EventArgs.X +
    '                startEvent.EventArgs.Y * startEvent.EventArgs.Y) > MinimumOffset
    '          Select startEvent).TimeInterval
    '        Where interval.Interval.TotalMilliseconds < TimeThreshold
    '        Select New Unit

    '    Return query

    'End Function
End Module

File: Dice/RxWp7Dice/Dice/ShakeObserver.cs
using System;
using System.Net;
using System.Windows;
using System.Windows.Controls;
using System.Windows.Documents;
using System.Windows.Ink;
using System.Windows.Input;
using System.Windows.Media;
using System.Windows.Media.Animation;
using System.Windows.Shapes;

using System.Linq;
using Microsoft.Devices.Sensors;
using Microsoft.Phone.Reactive;
using System.Collections;
using System.Collections.Generic;

namespace Dice
{
    public class ShakeObserver
    {
        const double MinimumOffset = 1.44;
        const double TimeThreshold = 200;
        public static IObservable<IEvent<AccelerometerReadingEventArgs>> GetShakeObserver(Accelerometer accel)
        {
            var query =  
                from interval in 
                (from startEvent in Observable.FromEvent<AccelerometerReadingEventArgs>(accel, "ReadingChanged")
                   where startEvent.EventArgs.X*startEvent.EventArgs.X + 
                        startEvent.EventArgs.Y*startEvent.EventArgs.Y > MinimumOffset
                   select startEvent).TimeInterval()
                   where interval.Interval.TotalMilliseconds < TimeThreshold 
                   select interval.Value;

            return query;
        }
    }

    public class foo<T> : IEnumerator<T>, IObserver<T>
    {
        public T Current
        {
            get { throw new NotImplementedException(); }
        }

        public void Dispose()
        {
            throw new NotImplementedException();
        }

        object IEnumerator.Current
        {
            get { throw new NotImplementedException(); }
        }

        public bool MoveNext()
        {
            throw new NotImplementedException();
        }

        public void Reset()
        {
            throw new NotImplementedException();
        }

        public void OnCompleted()
        {
            throw new NotImplementedException();
        }

        public void OnError(Exception exception)
        {
            throw new NotImplementedException();
        }

        public void OnNext(T value)
        {
            throw new NotImplementedException();
        }
    }

}
